contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,410 @@
+  * I want a consistent and portable solution to consuming hardware devices
+    across k8s clusters
+
+## Objectives
+
+1. Create a plugin mechanism which allows discovery and monitoring of devices
+2. Add support for Devices in kubelet
+3. Define a deployment mechanism
+4. Versioning the new interface/API
+
+## Non Objectives
+1. Metrics this should be the job of cadvisor and should probably either be
+   addressed there (cadvisor) or if people feel there is a case to be made
+   for it being addressed in the Device Plugin in a follow up proposal.
+
+## Installation
+
+The installation process should be straightforward to the user, transparent
+and similar to other regular Kubernetes actions.
+The device plugin should also run in containers so that kubernetes can
+deploy them and restart the plugins when they fail.
+However, we should not prevent the user from deploying a bare metal device
+plugin.
+
+Deploying the device plugins though DemonSets makes sense as the cluster
+admin would be able to specify which machines it wants the device plugins to
+run on, the process is similar to any kubernetes action and does not require
+to change any parts of Kubernetes.
+
+Additionally, for integrated solutions such as `kubeadm` we can add support
+to auto-deploy community vetted Device Plugins.
+Thus not fragmenting once more the Kubernetes ecosystem.
+
+For users installing Kubernetes without using an integrated solution such
+as `kubeadm` they would use the examples that we would provide at:
+`https://github.com/kubernetes/kubernetes/tree/master/examples/device-plugin.yaml`
+
+YAML example:
+```yaml
+apiVersion: extensions/v1beta1
+kind: DaemonSet
+metadata:
+spec:
+    template:
+        metadata:
+            labels:
+                - name: device-plugin
+        spec:
+            containers:
+                name: device-plugin-ctr
+                image: nvidia/device-plugin:1.0
+                volumeMounts:
+                  - mountPath: /device-plugin
+                  - name: device-plugin
+           volumes:
+             - name: device-plugin
+               hostPath:
+                   path: /var/run/kubernetes
+```
+
+## API Changes
+### Device
+
+When discovering the devices, Kubelet will be in charge of advertising those
+resources to the API server.
+
+We will advertise each device returned by the Device Plugin in a new structure
+called `Device`.
+It is defined as follows:
+
+```golang
+type Device struct {
+	Kind       string
+	Vendor     string
+	Name       string
+	Health     DeviceHealthStatus
+	Properties map[string]string
+}
+```
+
+Because the current API (Capacity) can not be extended to support Device,
+we will need to create two new attributes in the NodeStatus structure:
+  * `DevCapacity`: Describing the device capacity of the node
+  * `DevAvailable`: Describing the available devices
+
+```golang
+type NodeStatus struct {
+	DevCapacity []Device
+	DevAvailable []Device
```

Comment:
As mentioned on the previous comment Allocatable != Available: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node-allocatable.md

DevAvailable should not be named DevAllocatable because it does not do the same thing